src/javascript.grammar

```diff
@@ -31,15 +22,21 @@
   ternary @right,
   assign @right,
   comma @left,
-  statement @cut,
-  predicate
+  cellname @cut,
+  statement @cut
 }
 
-@top Script { statement+ }
+@top Cell {
+  ImportDeclaration |
+  Block |
+  ExpressionStatement |
+  NamedBlockCell[@dynamicPrecedence=1] { (VariableDefinition | QualifiedVariableDefinition) !cellname "=" Block } |
+  NamedExpressionCell[@dynamicPrecedence=1] { (VariableDefinition | QualifiedVariableDefinition) !cellname "=" expression semi } |
+  NamedClassCell { kw<"class"> VariableDefinition !cellname (kw<"extends"> expression)? ClassBody } |
+  NamedFunctionCell { async? kw<"function"> Star? VariableDefinition !cellname ParamList Block }
```

Comment:
just to make sure I'm understanding this correctly, the `!cellname` precedence marker here so that `NamedClassCell` and `NamedFunctionCell` are always matched instead of `ClassExpression` and `AssignmentExpression`?

Also maybe it makes sense that `@dynamicPrecedence=1` is needed for `NamedExpressionCell` bc the `assign @right` marker takes precedence over the `cellname @cut` one (bc the former is listed higher than the latter) and they're ambiguous? idk if that would make a difference... also though seems odd that it's needed for `NamedBlockCell` because that doesn't match `AssignmentExpression`, does it? since a `Block` isn't an `expressionNoComma`? 🤔

Reply:
My intent with the `!cellname` and `@precedence { cellname @cut }` was that as soon as the parser sees a cell starting with `class Foo`, it knows that it’s a NamedClassCell and doesn’t need to follow the ExpressionStatement path (which could be a ClassExpression). And then similarly for NamedFunctionCell over FunctionExpression, NamedExpressionCell over AssignmentExpression, and NamedBlockCell over AssignmentExpression + ObjectExpression (ambiguous with Block).

I found that this worked as expected as long as NamedBlockCell etc. were using VariableName, but as soon as I switched to VariableDefinition, the precedence operator no longer worked as expected, and I had to add the `@dynamicPrecedence=1`.

Let me try tweaking a few things and seeing if the tests pass…

Reply:
Yeah, I can’t get it to work without `[@dynamicPrecedence=1]`… and further more, the `!cellname` precedence operator doesn’t seem to have any effect. I suppose I could just take that out, but my hope is that it makes the parser more efficient. No idea why it’s not working though.